### PR TITLE
EVG-14802: Implement DeletePod and test ECSPod and ECSPodCreator

### PR DESCRIPTION
**Pull request by jessfeng**
[EVG-14802](https://jira.mongodb.org/browse/EVG-14802)

**Description**
- Implement DeletePod
- Change `Tags` structure to a map instead of slice

**Testing**
(bulk of the code added is here)
- `ECSPodCreator`- creating pod creator and creating pods
- `ECSPod `- stopping and deleting pods (shared tests), info checking and creating a basic pod with invalid pod options are in the ecs specific `pod_test` file

### CODE REVIEW

File: ecs/pod_creator_test.go [hunk 1]
```diff
@@ -1,12 +1,286 @@
...
+
+			defer func() {
+				require.NoError(t, p.Delete(ctx))
+			}()
+
+			info, err := p.Info(ctx)
+			require.NoError(t, err)
+			assert.Equal(t, cocoa.Running, info.Status)
+		},
+		"CreatePodSucceedsWithEVNoVault": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:209]: "WithEnvironmentVariables" or "WithEnvVars"

File: ecs/pod_creator_test.go [hunk 2]
```diff
@@ -1,12 +1,286 @@
...
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container").
+				SetMemoryMB(128).
+				SetCPU(128)
+			require.NotNil(t, containerDef.EnvVars)
+
+			execOpts := cocoa.NewECSPodExecutionOptions().
+				SetCluster(testutil.ECSClusterName()).
+				AddTags(map[string]string{"k0": "v0", "k1": "v1"})
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:220]: Remove tags, since I don't think they're relevant to this test.

File: ecs/pod_creator_test.go [hunk 3]
```diff
@@ -1,12 +1,286 @@
...
+
+			info, err := p.Info(ctx)
+			require.NoError(t, err)
+			assert.Equal(t, cocoa.Running, info.Status)
+		},
+		"CreatePodSucceedsWithEVNoVault": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("name").SetValue("value")
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:213]: Is it necessary to set the name or can we omit it?

File: internal/testcase/ecs_pod.go [hunk 4]
```diff
@@ -1 +1,159 @@
...
+	envVar := cocoa.NewEnvironmentVariable().SetName("name").SetValue("value")
+	secret := cocoa.NewEnvironmentVariable().SetName("secret1").
+		SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name1")).SetValue("value1"))
+	secretOwned := cocoa.NewEnvironmentVariable().SetName("secret2").
+		SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name2")).SetValue("value2"))
+	secretOwned.SecretOpts.SetOwned(true)
+
+	containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+		SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+		SetName("container").
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:27]: I would omit this if it's not required and not relevant to the test .

File: ecs/pod_creator_test.go [hunk 5]
```diff
@@ -1,12 +1,286 @@
...
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodFailsWithIncompleteContainerDef": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name")).SetValue("value"))
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container")
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:80]: Is it necessary to set the name or can we omit it?

File: ecs/pod_creator_test.go [hunk 6]
```diff
@@ -1,12 +1,286 @@
...
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodFailsWithInvalidSecretName": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName("name").SetValue("value"))
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:109]: Is it necessary to set the name or can we omit it?

File: ecs/pod_creator_test.go [hunk 7]
```diff
@@ -1,12 +1,286 @@
...
+
+			execOpts := cocoa.NewECSPodExecutionOptions().SetCluster(testutil.ECSClusterName())
+			assert.NoError(t, execOpts.Validate())
+
+			opts := cocoa.NewECSPodCreationOptions().SetName("name").
+				AddContainerDefinitions(*containerDef).
+				SetMemoryMB(128).
+				SetCPU(128).
+				SetTaskRole("role").
+				AddTags("tag").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:122]: Tags don't seem relevant to this test and they're not a required parameter, so I would remove them.

File: ecs/pod_creator_test.go [hunk 8]
```diff
@@ -1,12 +1,286 @@
...
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName("name").SetValue("value"))
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container").
+				SetMemoryMB(128).
+				SetCPU(128)
+			require.NotNil(t, containerDef.EnvVars)
+
+			execOpts := cocoa.NewECSPodExecutionOptions().SetCluster(testutil.ECSClusterName())
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:114]: This test should set the execution role to ensure that the test doesn't just fail because the execution role is missing.

File: ecs/pod_creator_test.go [hunk 9]
```diff
@@ -1,12 +1,286 @@
...
+
+			podCreator, err := NewBasicECSPodCreator(c, nil)
+			require.NoError(t, err)
+
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodFailsWithIncompleteContainerDef": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:76]: Remove the environment variable. It's not relevant to this test.

File: ecs/pod_creator_test.go [hunk 10]
```diff
@@ -1,12 +1,286 @@
...
+			require.Zero(t, p)
+		},
+		"CreatePodFailsWithSecretsNoVault": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("evName").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName("name").SetValue("value"))
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar})
+			require.NotNil(t, containerDef.EnvVars)
+
+			execOpts := cocoa.NewECSPodExecutionOptions().SetCluster(testutil.ECSClusterName())
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:54]: This test should set the execution role to ensure that the test doesn't just fail because the execution role is missing.

File: ecs/pod_creator_test.go [hunk 11]
```diff
@@ -1,12 +1,286 @@
...
+			assert.NoError(t, opts.Validate())
+
+			podCreator, err := NewBasicECSPodCreator(c, v)
+			require.NoError(t, err)
+
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodFailsWithInvalidSecretName": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:104]: What exactly is invalid about the secret name? I'm pretty sure the options validation in the pod creator would prevent it from being an invalid name. Do you mean to actually test the condition "the secret doesn't exist in Secrets Manager, but we also don't create it"?

File: ecs/pod_creator_test.go [hunk 12]
```diff
@@ -1,12 +1,286 @@
...
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodFailsWithSecretsButNoExecutionRole": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name")).SetValue("value"))
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:140]: Is it necessary to set the name or can we omit it?

File: ecs/pod_creator_test.go [hunk 13]
```diff
@@ -1,12 +1,286 @@
...
+			podCreator, err := NewBasicECSPodCreator(c, v)
+			require.NoError(t, err)
+
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodSucceeds": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name")).SetValue("value"))
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:169]: I imagine this would have to call `SetExists(false)` to actually create the secret and `SetOwned(true)` to properly clean it up.

File: ecs/pod_creator_test.go [hunk 14]
```diff
@@ -1,12 +1,286 @@
...
+			assert.NoError(t, execOpts.Validate())
+
+			opts := cocoa.NewECSPodCreationOptions().
+				AddContainerDefinitions(*containerDef).
+				SetMemoryMB(128).
+				SetCPU(128).
+				SetTaskRole(testutil.TaskRole()).
+				AddTags("tag").
+				SetExecutionOptions(*execOpts)
+			require.NotZero(t, opts)
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:189]: You don't need to check L189-190.

File: ecs/pod_creator_test.go [hunk 15]
```diff
@@ -1,12 +1,286 @@
...
+			execOpts := cocoa.NewECSPodExecutionOptions().
+				SetCluster(testutil.ECSClusterName()).
+				AddTags(map[string]string{"k0": "v0", "k1": "v1"})
+			assert.NoError(t, execOpts.Validate())
+
+			opts := cocoa.NewECSPodCreationOptions().
+				AddContainerDefinitions(*containerDef).
+				SetMemoryMB(128).
+				SetCPU(128).
+				SetTaskRole(testutil.TaskRole()).
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:227]: I think `SetTaskRole` is unnecesary if it's not using any secrets.

File: ecs/pod_creator_test.go [hunk 16]
```diff
@@ -1,12 +1,286 @@
...
+				SetExecutionOptions(*execOpts)
+			require.NotZero(t, opts)
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:229]: You don't need to check L229-230.

File: ecs/pod_creator_test.go [hunk 17]
```diff
@@ -1,12 +1,286 @@
...
+
+	for tName, tCase := range map[string]func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault){
+		"NewPodCreatorFailsWithMissingClient": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			podCreator, err := NewBasicECSPodCreator(nil, nil)
+			require.Error(t, err)
+			require.Zero(t, podCreator)
+		},
+		"CreatePodFailsWithInvalidCreationOpts": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			opts := cocoa.NewECSPodCreationOptions()
+			require.NotZero(t, opts)
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:37]: You don't need to check L37-38 since the options will never return nil.

File: ecs/pod_creator_test.go [hunk 18]
```diff
@@ -1,12 +1,286 @@
...
+			assert.NoError(t, execOpts.Validate())
+
+			opts := cocoa.NewECSPodCreationOptions().SetName("name").
+				AddContainerDefinitions(*containerDef).
+				SetMemoryMB(128).
+				SetCPU(128).
+				SetTaskRole("role").
+				AddTags("tag").
+				SetExecutionOptions(*execOpts)
+			require.NotZero(t, opts)
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:124]: Same as above - you don't need to check for non-nil options.

File: ecs/pod_creator_test.go [hunk 19]
```diff
@@ -1,12 +1,286 @@
...
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodSucceeds": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().SetName("ev").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name")).SetValue("value"))
+			containerDef := cocoa.NewECSContainerDefinition().SetImage("image").
+				SetEnvironmentVariables([]cocoa.EnvironmentVariable{*envVar}).
+				SetName("container").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:172]: I would omit the container name if it's not required and irrelevant to the test.

File: ecs/pod_creator_test.go [hunk 20]
```diff
@@ -1,12 +1,286 @@
...
+				SetCluster(testutil.ECSClusterName()).
+				SetExecutionRole(testutil.ExecutionRole())
+			assert.NoError(t, execOpts.Validate())
+
+			opts := cocoa.NewECSPodCreationOptions().
+				AddContainerDefinitions(*containerDef).
+				SetMemoryMB(128).
+				SetCPU(128).
+				SetTaskRole(testutil.TaskRole()).
+				AddTags("tag").
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:187]: I would omit this if it's not required and irrelevant to the test.

File: ecs/pod_creator_test.go [hunk 21]
```diff
@@ -1,12 +1,286 @@
...
+			assert.NoError(t, opts.Validate())
+
+			podCreator, err := NewBasicECSPodCreator(c, v)
+			require.NoError(t, err)
+
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodSucceeds": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:167]: "CreatePodSucceedsWithNewlyCreatedSecrets"

File: ecs/pod_test.go [hunk 22]
```diff
@@ -54,9 +63,61 @@ func TestECSPod(t *testing.T) {
...
+			defer c.Close(ctx)
 
 			tCase(tctx, t, c)
 		})
+
+	}
+
+}
+
+func TestECSPod(t *testing.T) {
```
Comment (Kimchelly) [inline ecs/pod_test.go:75]: I don't think this is worth doing now, but can you make a ticket and attach a TODO comment to clean up the resources in this test more thoroughly? Since this test has the potential to leak resources (secrets, task definitions, and tasks), we need to clean up to make sure that after all the tests have finished running, there's no lingering resources in AWS. I would make a comment like this:
`// TODO (<TICKET_NUMBER>): clean up resources in ECS tests more thoroughly.

File: internal/testcase/ecs_pod.go [hunk 23]
```diff
@@ -1 +1,159 @@
...
+	execOpts := cocoa.NewECSPodExecutionOptions().
+		SetCluster(testutil.ECSClusterName()).
+		SetExecutionRole(testutil.ExecutionRole())
+
+	opts := cocoa.NewECSPodCreationOptions().
+		AddContainerDefinitions(*containerDef).
+		SetMemoryMB(128).
+		SetCPU(128).
+		SetTaskRole(testutil.TaskRole()).
+		AddTags("tag").
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:40]: I would omit this if it's not required and not relevant to the test.

File: internal/testcase/ecs_pod.go [hunk 24]
```diff
@@ -1 +1,159 @@
...
+		SetExecutionOptions(*execOpts)
+
+	optsSecret := cocoa.NewECSPodCreationOptions().
+		AddContainerDefinitions(*containerDef.AddEnvironmentVariables(*secret, *secretOwned)).
+		SetMemoryMB(128).
+		SetCPU(128).
+		SetTaskRole(testutil.TaskRole()).
+		AddTags("tag").
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:48]: I would omit this if it's not required and not relevant to the test .

File: internal/testcase/ecs_pod.go [hunk 25]
```diff
@@ -1 +1,159 @@
...
+		},
+		"StopSucceedsWithSecrets": func(ctx context.Context, t *testing.T, v cocoa.Vault, pc cocoa.ECSPodCreator) {
+			p, err := pc.CreatePod(ctx, optsSecret)
+			require.NoError(t, err)
+			require.NotZero(t, p)
+
+			info, err := p.Info(ctx)
+			require.NoError(t, err)
+			require.NotNil(t, info)
+			require.NotNil(t, info.Resources)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:74]: This is trivially true because `info.Resources` is not a pointer. If you want to check if it has any values set, you can do `require.NotZero(t, info.Resources)`.

File: internal/testcase/ecs_pod.go [hunk 26]
```diff
@@ -1 +1,159 @@
...
+			assert.Equal(t, cocoa.Running, info.Status)
+			assert.Equal(t, 2, len(info.Resources.Secrets))
+
+			err = p.Stop(ctx)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:78]: Combine L78-79.

File: internal/testcase/ecs_pod.go [hunk 27]
```diff
@@ -1 +1,159 @@
...
+			id, err := v.GetValue(ctx, *arn)
+			require.NoError(t, err)
+			require.NotNil(t, id)
+		},
+		"StopFailsOnIncorrectPodStatus": func(ctx context.Context, t *testing.T, v cocoa.Vault, pc cocoa.ECSPodCreator) {
+			p, err := pc.CreatePod(ctx, opts)
+			require.NoError(t, err)
+			require.NotZero(t, p)
+
+			err = p.Stop(ctx)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:97]: Combine L97-98.

File: internal/testcase/ecs_pod.go [hunk 28]
```diff
@@ -1 +1,159 @@
...
+		},
+		"DeleteSucceedsWithSecrets": func(ctx context.Context, t *testing.T, v cocoa.Vault, pc cocoa.ECSPodCreator) {
+			p, err := pc.CreatePod(ctx, optsSecret)
+			require.NoError(t, err)
+			require.NotZero(t, p)
+
+			info, err := p.Info(ctx)
+			require.NoError(t, err)
+			require.NotNil(t, info)
+			require.NotNil(t, info.Resources)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:134]: This is trivially true.

File: internal/testcase/ecs_pod.go [hunk 29]
```diff
@@ -1 +1,159 @@
...
+			require.NotNil(t, info.Resources.Secrets)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:135]: You probably want to check `require.NotEmpty(t, info.Resources.Secrets)` because the slice could be empty. However, I would instead just remove this check since L137 already checks that the secrets are both non-empty and of length 2.

File: internal/testcase/ecs_pod.go [hunk 30]
```diff
@@ -1 +1,159 @@
...
+			assert.Equal(t, cocoa.Running, info.Status)
+			require.Equal(t, 2, len(info.Resources.Secrets))
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:137]: You can just do `require.Len(t, info.Resources.Secrets, 2)`.

File: internal/testcase/ecs_pod.go [hunk 31]
```diff
@@ -1 +1,159 @@
...
+
+			err = p.Delete(ctx)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:139]: Combine L139-140.

File: internal/testcase/ecs_pod.go [hunk 32]
```diff
@@ -1 +1,159 @@
...
+			require.NoError(t, err)
+
+			info, err = p.Info(ctx)
+			require.NoError(t, err)
+			require.NotNil(t, info)
+			assert.Equal(t, cocoa.Deleted, info.Status)
+			require.Equal(t, 2, len(info.Resources.Secrets))
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:146]: Same as above. `require.Len(t, info.Resources.Secrets, 2)`

File: internal/testcase/ecs_pod.go [hunk 33]
```diff
@@ -1 +1,159 @@
...
+
+			arn0 := info.Resources.Secrets[0].Name
+			id, err := v.GetValue(ctx, *arn0)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:149]: This is returning the value, not the ID. I would rename `id` to `val`.

File: internal/testcase/ecs_pod.go [hunk 34]
```diff
@@ -1 +1,159 @@
...
+			require.NoError(t, err)
+			require.NotZero(t, id)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:151]: Is this supposed to check that the returned value equals the one you originally set?

**Review by Kimchelly — CHANGES_REQUESTED** [review]

File: ecs/pod_creator_test.go (hunk 5 above)
Comment (jessfeng) [reply ecs/pod_creator_test.go:80]: I set the name because otherwise it throws an error (`ClientException: Container.name should not be null or empty.`) when making an API call to register the task definition. Instead of explicitly setting a name, I'll just require no error when validating the container definition because the validation function auto-generates a name if it's nil.

File: ecs/pod_test.go (hunk 22 above)
Comment (jessfeng) [reply ecs/pod_test.go:75]: Sure, done.

File: internal/testcase/ecs_pod.go (hunk 4 above)
Comment (jessfeng) [reply internal/testcase/ecs_pod.go:27]: This is required when registering the task definition--should I keep it as it is (`SetName`) or run `containerDef.Validate()` to have the validation function generate a name for the container? My hesitancy with the latter is that we would have to check the error, but there wouldn't be one. Another thing I could do is set the name with `utility.RandomString()`.

Comment (Kimchelly) [reply internal/testcase/ecs_pod.go:27]: If it's a required parameter, then it's okay to leave it.

File: ecs/pod_creator_test.go (hunk 5 above)
Comment (Kimchelly) [reply ecs/pod_creator_test.go:80]: Same as above - it's okay to keep.

File: ecs/pod_creator_test.go (hunk 11 above)
Comment (jessfeng) [reply ecs/pod_creator_test.go:104]: I don't recall the point of this test, so I'm just going to remove it. If the secret doesn't exist in SM and we don't create it, is that when we declare a secret and its fields but the container definition doesn't include it in its environment variable list?

Comment (Kimchelly) [reply ecs/pod_creator_test.go]: Removing the test SGTM.

File: ecs/pod_creator_test.go [hunk 35]
```diff
@@ -1,12 +1,249 @@
...
+
+			p, err := podCreator.CreatePod(ctx, opts)
+			require.Error(t, err)
+			require.Zero(t, p)
+		},
+		"CreatePodSucceedsWithNewlyCreatedSecrets": func(ctx context.Context, t *testing.T, c cocoa.ECSClient, v cocoa.Vault) {
+			envVar := cocoa.NewEnvironmentVariable().
+				SetName("envVar").
+				SetSecretOptions(*cocoa.NewSecretOptions().SetName(testutil.NewSecretName("name")).SetValue("value"))
+			envVar.SecretOpts.SetExists(false)
```
Comment (Kimchelly) [inline ecs/pod_creator_test.go:134]: Nit: can you create the secret options like this?
```go
cocoa.NewSecretOptions().
    SetName(...).
    SetValue(...).
    SetExists(false).
    SetOwned(true)
```
The method spacing is optional, but it helps readability if you chain several method calls and put them on separate lines.

File: internal/testcase/ecs_pod.go [hunk 36]
```diff
@@ -1 +1,165 @@
...
+			require.NoError(t, p.Delete(ctx))
+
+			info, err = p.Info(ctx)
+			require.NoError(t, err)
+			require.NotNil(t, info)
+			assert.Equal(t, cocoa.Deleted, info.Status)
+			require.Len(t, info.Resources.Secrets, 2)
+
+			arn0 := info.Resources.Secrets[0].Name
+			val, err := v.GetValue(ctx, *arn0)
```
Comment (Kimchelly) [inline internal/testcase/ecs_pod.go:155]: Just to check my understanding of this test - even if you've already deleted the secret with `ForceDeleteWithoutRecovery`, it'll still return the value to you?

**Review by Kimchelly — APPROVED** [review]
LGTM mod nit and question.

File: ecs/pod_creator_test.go (hunk 35 above)
Comment (jessfeng) [reply ecs/pod_creator_test.go:134]: Yep, but I called `SetOwned` on a separate line since it returns a `PodSecret` type, which isn't compatible with the `SecretOpts` type.

File: internal/testcase/ecs_pod.go (hunk 36 above)
Comment (jessfeng) [reply internal/testcase/ecs_pod.go:155]: No, it shouldn't because the secret is already scheduled for deletion. However, in this case, the first secret (`secret`) is not owned by the pod, so it doesn't get deleted, which is why I check that there is no error and the value that getting the secret's value returns us is the one we set it as in the beginning. For the second secret (`secretOwned`), since its `Owned` field is set as true, it does get deleted and returns an error when we try to get its value.
